Lib/gftools/builder/dependencies.py

```diff
@@ -0,0 +1,173 @@
+            for dist in importlib_metadata.distributions()
+        )
+        if name not in excluded
+    }
+
+
+def get_provided_extras(dist_name) -> Set[str]:
+    """Return the set of extras provided by the given distribution package."""
+    metadata = importlib_metadata.metadata(dist_name)
+    if "Provides-Extra" in metadata:
+        return {_normalize(e) for e in metadata.get_all("Provides-Extra")}
+    return set()
+
+
+def get_installed_requirements(
+    dist_name: str,
+    extras: Set[str] = frozenset(),
+    excluded: Set[str] = DENYLIST,
+) -> Generator[Tuple[str, str], None, None]:
+    """Yield (name, version) of a distribution package and all its dependencies.
+
+    Parse the installed packages' metadata, and traverse the dependency graph starting
+    at 'dist_name'. Skip packages that are not installed, explicitly excluded (e.g.
+    setuptools), or which don't match the requested set of 'extras'.
+    """
+    installed = get_installed_distributions(excluded)
+
+    visited: DefaultDict[str, Set[str]] = defaultdict(set)
+    frontier = deque([(_normalize(dist_name), extras)])
+
+    while frontier:
+        dist_name, extras = frontier.popleft()
+        if dist_name not in installed:
+            continue
+
+        dist = installed[dist_name]
+
+        if dist_name not in visited:
+            yield (dist_name, dist.version)
+
+        for req_spec in dist.requires or ():  # can be None
+            req = Requirement(req_spec)
+            # skip optional dependencies that don't match the requested extras
+            if not _evaluate_extras(req, extras):
+                continue
+            req_name = _normalize(req.name)
+            req_extras = {_normalize(extra) for extra in req.extras}
+
+            if req_name not in visited:
+                frontier.append((req_name, req_extras))
+            else:
+                # readd to the queue to traverse unvisited optional branches
+                unvisited = req_extras - visited[req_name]
+                if unvisited:
+                    frontier.append((req_name, unvisited))
+
+        visited[dist_name] |= extras
+
+
+def format_requirements(requirements: Iterable[Tuple[str, str]]) -> str:
+    return "".join(f"{name}=={version}\n" for name, version in sorted(requirements))
+
+
+def build_font_requirements(ttfont: TTFont, dist_name: str):
+    if "Debg" in ttfont:
+        debg = ttfont["Debg"]
+    else:
+        debg = newTable("Debg")
+        debg.data = {}
+    requirements = get_installed_requirements(dist_name)
+    debg.data["com.github.googlefonts.gftools.deps"] = format_requirements(
+        requirements
+    ).splitlines()
+    ttfont["Debg"] = debg
+
+
+def dump_font_requirements(ttfont: TTFont):
+    assert "Debg" in ttfont, "No Debg table found!"
+    debg_table = ttfont["Debg"].data
+    assert (
+        GFTOOLS_DEPENDENCIES_KEY in debg_table
+    ), f"Missing '{GFTOOLS_DEPENDENCIES_KEY}' in Debg table"
```

Comment:
the original KeyError would be better than an AssertionError. Assertions are supposed to never trigger in normal circumstances, whereas here one may well attempt to dump deps from a font that has none and an exception is the right approach in this case. If you don't like KeyError and want to customize the error message you can reraise it as whatever other exception you prefer